runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -11963,3 +11986,295 @@ J9::Z::TreeEvaluator::tabortEvaluator(TR::Node * node, TR::CodeGenerator * cg)
+   startICFLabel->setStartInternalControlFlow();
+
+   int32_t shiftAmount = TR::Compiler->om.compressedReferenceShift();
+   bool shouldShift = (shiftAmount != 0);
+   TR::InstOpCode::Mnemonic loadOpCode = isCompressedRef ? TR::InstOpCode::LLGF: TR::InstOpCode::LG;
+
+   if (fieldUnresolved)
+      {
+      generateRXInstruction(cg, TR::InstOpCode::LAY, node, locationReg, loadMemRef);
+      generateRXInstruction(cg, loadOpCode, node, resultReg, generateS390MemoryReference(locationReg, 0, cg));
+      }
+   else
+      {
+      generateRXInstruction(cg, loadOpCode, node, resultReg, loadMemRef);
+      }
+   if (shouldShift)
+      {
+      generateRSInstruction(cg, TR::InstOpCode::SLLG, node, resultReg, resultReg, shiftAmount);
+      }
+
+   if (loadMemRef->getIndexRegister() != NULL)
+      {
+      deps->addPostCondition(loadMemRef->getIndexRegister(), TR::RealRegister::AssignAny);
+      }
+   if (loadMemRef->getBaseRegister() != NULL)
+      {
+      deps->addPostCondition(loadMemRef->getBaseRegister(), TR::RealRegister::AssignAny);
+      }
+
+   TR::Register* vmReg = cg->getLinkage()->getMethodMetaDataRealRegister();
+   TR::MemoryReference* baseMemRef = generateS390MemoryReference(vmReg, TR::Compiler->vm.thisThreadGetEvacuateBaseAddressOffset(comp), cg);
+   TR::MemoryReference* topMemRef  = generateS390MemoryReference(vmReg, TR::Compiler->vm.thisThreadGetEvacuateTopAddressOffset(comp), cg);
+
+   // Range check with implicit software CS status check.
+   generateRXInstruction(cg, TR::InstOpCode::CLG, node, resultReg, baseMemRef);
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BL, node, doneLabel);
+   generateRXInstruction(cg, TR::InstOpCode::CLG, node, resultReg, topMemRef);
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BH, node, doneLabel);
+
+   if (!fieldUnresolved)
+      {
+      generateRXInstruction(cg, TR::InstOpCode::LAY, node, locationReg, generateS390MemoryReference(*loadMemRef, 0, cg));
+      }
+   // Save the object address to vmThread->floatTemp1 for jitReadBarrier helper.
+   generateRXInstruction(cg, TR::InstOpCode::STG, node, locationReg, generateS390MemoryReference(vmReg, offsetof(J9VMThread, floatTemp1), cg));
+
+   cg->generateDebugCounter(TR::DebugCounter::debugCounterName(cg->comp(), "readBar/helperCall"), 1, TR::DebugCounter::Cheap);
+   TR::LabelSymbol* callLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol* callEndLabel = generateLabelSymbol(cg);
+   TR::Instruction *gcPoint = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, callLabel);
+   gcPoint->setNeedsGCMap(0);
+   auto readBarHelperSnippet = new (cg->trHeapMemory()) TR::S390HelperCallSnippet(cg, node, callLabel,
+                                                                     cg->symRefTab()->findOrCreateRuntimeHelper(TR_readBarrier, false, false, false),
+                                                                     callEndLabel);
```

Comment:
This helper call seems super hacky. We store a value to `floatTemp1` and then re-load it in the helper? Why aren't we enforcing proper linkage parameter passing? Can you touch base with @dchopra001 as he recently implemented a `TR_Helper` call with proper linkage to see how it works. We don't need to go via a snippet call (`TR::S390HelperCallSnippet`) at all.

In fact this is making me wonder whether we should just bit the bullet and implement a proper `TR_Helper` linkage (outside the scope of this PR).

Reply:
I changed it back to use `vm.thisThreadGetGSOperandAddressOffset` and `vm.thisThreadGetGSHandlerAddressOffset` so the SW and HW implementations on z are consistent in terms of helper calls.